### PR TITLE
 Apacdex : add new adapter and remove Valueimpression adapter.

### PR DESCRIPTION
**Pull request by marki1an**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        return BidderBid.of(bid, getBidType(typeNode, bid.getImpid()), bidResponse.getCur());
+    }
+
+    private BidType getBidType(JsonNode typeNode, String impid) {
+        try {
+            return mapper.mapper().convertValue(typeNode, BidType.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(String.format("Failed to parse bid media type for impression %s", impid));
+        }
+    }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java:133]: I think better to do it like this:
```
    private BidderBid constructBidderBid(Bid bid, String currency, List<BidderError> errors) {
        final JsonNode extNode = bid.getExt();
        final JsonNode typeNode = isNotEmptyOrMissedNode(extNode) ? extNode.at("/prebid/type") : null;

        try {
            return BidderBid.of(bid, getBidType(typeNode, bid.getImpid()), currency);
        } catch (PreBidException e) {
            errors.add(BidderError.badInput(e.getMessage()));
            return null;
        }
    }

    private BidType getBidType(JsonNode typeNode, String impid) {
        try {
            return mapper.mapper().convertValue(typeNode, BidType.class);
        } catch (IllegalArgumentException ignored) {
        }

        throw new PreBidException(String.format("Failed to parse bid media type for impression %s", impid));
    }
```

File: src/main/resources/bidder-config/apacdex.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+adapters:
```
Comment (CTMBNara) [inline src/main/resources/bidder-config/apacdex.yaml:1]: Add `quantumdex` and `valueimpression` aliases.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java (hunk 1 above)
Comment (marki1an) [reply src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java:133]: we can omit `try{...} catch()` if we in `getBidType(...)` ignore exception

Comment (marki1an) [reply src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java:133]: i think we can leave `type Node` check on null

File: src/test/java/org/prebid/server/bidder/apacdex/ApacdexBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    @Test
+    public void makeBidsShouldReturnErrorWhenUnrecognisedType() throws JsonProcessingException {
+        // given
+        final ObjectNode givenExt = getExtUnknownTypeJsonNode();
+
+        final HttpCall<BidRequest> httpCall = givenHttpCall(
+                BidResponse.builder()
+                        .seatbid(givenSeatBid(
+                                givenBid("123", null, bidBuilder -> bidBuilder.ext(givenExt))))
+                        .build());
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/apacdex/ApacdexBidderTest.java:170]: Add additional imp to verify that errors will not affect correct bids.

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 4]
```diff
@@ -52,6 +52,12 @@ adapters.algorix.enabled=true
 adapters.algorix.endpoint=http://localhost:8090/algorix-exchange
 adapters.amx.enabled=true
 adapters.amx.endpoint=http://localhost:8090/amx-exchange
+adapters.apacdex.enabled=true
+adapters.apacdex.endpoint=http://localhost:8090/apacdex-exchange
+adapters.apacdex.aliases.quantumdex.enabled = true
+adapters.apacdex.aliases.quantumdex.endpoint = http://localhost:8090/quantumdex-exchange
+adapters.apacdex.aliases.valueimpression.enabled = true
+adapters.apacdex.aliases.valueimpression.endpoint = http://localhost:8090/valueimpression-exchange
```
Comment (CTMBNara) [inline src/test/resources/org/prebid/server/it/test-application.properties:60]: Remove spaces for consistency.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/resources/bidder-config/apacdex.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    aliases:
+      quantumdex:
+        usersync:
+          url: https://sync.quantumdex.io/usersync/pbs?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r=
+          redirect-url: /setuid?bidder=quantumdex&gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}
+          cookie-family-name: quantumdex
+      valueimpression:
+        usersync:
+          url: https://sync.quantumdex.io/usersync/pbs?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r=
+          redirect-url: /setuid?bidder=quantumdex&gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}
```
Comment (CTMBNara) [inline src/main/resources/bidder-config/apacdex.yaml:14]: Is `quantumdex ` need to be replaced with `valueimpression` here?

File: src/test/java/org/prebid/server/bidder/apacdex/ApacdexBidderTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                                givenBid("44454", xNative)))
+                        .build());
+
+        // when
+        final Result<List<BidderBid>> result = apacdexBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .containsExactlyInAnyOrder(
```
[the remark this answers is not in the document]
Comment (And1sS) [reply src/test/java/org/prebid/server/bidder/apacdex/ApacdexBidderTest.java:154]: `containsExcatly` adds ordering constraint, which is unnecery in this case.

File: src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    private BidType getBidType(JsonNode typeNode, String impid) {
+        try {
+            return mapper.mapper().convertValue(typeNode, BidType.class);
+        } catch (IllegalArgumentException ignore) {
+        }
+
+        throw new PreBidException(String.format("Failed to parse bid media type for impression %s", impid));
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java:149]: Can be placed inside catch block. It's more in project style.

**Review by And1sS — CHANGES_REQUESTED** [review]

File: src/main/resources/bidder-config/apacdex.yaml (hunk 5 above)
Comment (marki1an) [reply src/main/resources/bidder-config/apacdex.yaml:14]: I had the same question but in go `quantumdex` 
https://github.com/prebid/prebid-server/pull/2097/files#diff-e3ddc5c58f8470f02b08100357f822cdc75d2c0692d1d9ba2fdd608c739df905R19

Comment (CTMBNara) [reply src/main/resources/bidder-config/apacdex.yaml:14]: Then remove `url` and `redirect-url` properties from aliases.

File: src/main/resources/bidder-config/apacdex.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      site-media-types:
+        - banner
+        - video
+      supported-vendors:
+      vendor-id: 0
+    usersync:
+      url: https://sync.quantumdex.io/usersync/pbs?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r=
+      redirect-url: /setuid?bidder=apacdex&gdpr=0&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}&uid=[UID]
+      cookie-family-name: apacdex
+      type: redirect
```
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/resources/bidder-config/apacdex.yaml]: Add secondary `usersync` configuration. Primary - `iframe`. Secondary - `redirect`.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<Imp> imps = new ArrayList<>();
+
+        if (CollectionUtils.isEmpty(request.getImp())) {
+            return Result.withError(BidderError.badInput("No Imps in Bid Request"));
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java:54]: Redundant check. See [this.](https://github.com/prebid/prebid-server-java/blob/c46d06fdb5ecbd1f1184ffe6a7056fc1bfcc9e74/src/main/java/com/iab/openrtb/request/BidRequest.java#L35)

**Review by CTMBNara — CHANGES_REQUESTED** [review]

**Review by CTMBNara — DISMISSED** [review]

**Review by CTMBNara — DISMISSED** [review]

**Review by And1sS — DISMISSED** [review]

File: src/main/resources/bidder-config/apacdex.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      vendor-id: 0
+    usersync:
+      url: https://sync.quantumdex.io/usersync/pbs?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r=
+      redirect-url: /setuid?bidder=apacdex&gdpr=0&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}&uid=[UID]
+      cookie-family-name: apacdex
+      type: iframe
+      support-cors: false
+      secondary:
+        url: https://sync.quantumdex.io/getuid?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r=
+        redirect-url: /setuid?bidder=apacdex&gdpr=0&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}&uid=[UID]
```
Comment (SerhiiNahornyi) [inline src/main/resources/bidder-config/apacdex.yaml:30]: This one will be specific for every bidder and alias. Pls update

File: src/main/resources/bidder-config/apacdex.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+adapters:
+  apacdex:
```
Comment (SerhiiNahornyi) [inline src/main/resources/bidder-config/apacdex.yaml:2]: I don't see enabled property for bidder and aliases. Also compare with other bidders config, and add missing to this file

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 12]
```diff
@@ -52,6 +52,12 @@ adapters.algorix.enabled=true
 adapters.algorix.endpoint=http://localhost:8090/algorix-exchange
 adapters.amx.enabled=true
 adapters.amx.endpoint=http://localhost:8090/amx-exchange
+adapters.apacdex.enabled=true
+adapters.apacdex.endpoint=http://localhost:8090/apacdex-exchange
+adapters.apacdex.aliases.quantumdex.enabled=true
```
Comment (SerhiiNahornyi) [inline src/test/resources/org/prebid/server/it/test-application.properties:57]: You added aliases test properties, but you haven't add tests for them. Pls add

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**